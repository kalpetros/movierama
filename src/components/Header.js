import React, { useContext } from 'react';

import { MoviesContext } from '../store/MoviesContext';

export const Header = () => {
  const { query, setQuery } = useContext(MoviesContext);

  const handleChange = (e) => {
    setQuery(e.currentTarget.value);
  };

  let title = 'In Theaters';
  if (query.length > 0) {
    title = `Movies "${query}"`;
  }

  return (
    <header className="header">
      <div className="header__content">
        <div>
          <div>
            <h1>Welcome</h1>
          </div>
          <div>
            <input
              type="text"
              value={query}
              className="input"
              placeholder="Type here to search for your favorite movie..."
              onChange={handleChange}
            ></input>
          </div>
        </div>
      </div>
      <div className="header__title">
        <div>
          <h1>{title}</h1>
        </div>
      </div>
    </header>
  );
};
