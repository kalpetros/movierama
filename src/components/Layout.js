import React from 'react';
import PropTypes from 'prop-types';

export const Layout = ({ children }) => {
  return (
    <main className="layout">
      <section>{children}</section>
    </main>
  );
};

Layout.propTypes = {
  children: PropTypes.node.isRequired,
};
