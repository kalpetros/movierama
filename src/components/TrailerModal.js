import React, { useState } from 'react';
import { createPortal } from 'react-dom';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export const TrailerModal = (props) => {
  const [visible, setVisible] = useState(false);
  const { id: id } = props;

  const handleClick = () => {
    setVisible((v) => (v ? false : true));
  };

  let portal = null;

  if (visible) {
    portal = createPortal(
      <div className="modal-wrapper" onClick={handleClick}>
        <div className="modal">
          <iframe
            width="1280"
            height="720"
            src={`https://www.youtube.com/embed/${id}`}
            frameBorder="0"
            allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
            allowFullScreen
          ></iframe>
        </div>
      </div>,
      document.getElementById('modal')
    );
  }

  return (
    <div>
      <FontAwesomeIcon
        icon={['fab', 'youtube']}
        size="2x"
        color="red"
        className="button-icon"
        onClick={handleClick}
      />
      {portal}
    </div>
  );
};

TrailerModal.propTypes = {
  id: PropTypes.string.isRequired,
};
