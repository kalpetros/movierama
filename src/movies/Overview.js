import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

export const Overview = (props) => {
  const [data, setData] = useState([]);
  const { id: id } = props;

  useEffect(() => {
    const fetchDetails = () => {
      const apiKey = 'bc50218d91157b1ba4f142ef7baaa6a0';
      const endpoint = `https://api.themoviedb.org/3/movie/${id}?api_key=${apiKey}`;

      fetch(endpoint, {
        method: 'GET',
      })
        .then((response) => {
          if (response.status !== 200) throw new Error(response.status);
          return response.json();
        })
        .then((data) => {
          setData(data);
        })
        .catch((error) => {
          console.log(error);
        });
    };

    fetchDetails();
  }, [id]);

  if (Object.keys(data).length === 0) {
    return null;
  }

  const overview = data.overview;
  const genres = data.genres.map((genre, index) => {
    return genre.name;
  });
  const video = data.video;
  const genresText = genres.join(', ');

  const producers = data.production_companies.map((company, index) => {
    return company.name;
  });

  const producersText = producers.length > 0 ? producers.join(', ') : 'N/A';

  const countries = data.production_countries.map((country, index) => {
    return country.name;
  });

  const countriesText = countries.length > 0 ? countries.join(', ') : 'N/A';

  const runtime = `${data.runtime} minutes`;

  return (
    <div className="movie__details__meta">
      <small className="font-bold text-uppercase">{genresText}</small>
      <p>{overview}</p>
      <h3>Runtime</h3>
      <small className="font-bold text-uppercase">{runtime}</small>
      <h3>Producers</h3>
      <small className="font-bold text-uppercase">{producersText}</small>
      <h3>Countries</h3>
      <small className="font-bold text-uppercase">{countriesText}</small>
    </div>
  );
};

Overview.defaultProps = {
  id: 1,
};

Overview.propTypes = {
  id: PropTypes.number.isRequired,
};
