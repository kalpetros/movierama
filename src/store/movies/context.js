import React, { createContext, useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';

import { buildEndpoint } from './utils';
import { fetchMovies } from './utils';

export const MoviesContext = createContext([]);

export const MoviesStateProvider = ({ children }) => {
  const [data, setData] = useState({
    init: true,
    page: 1,
    pages: 0,
    results: [],
  });
  const [query, setQuery] = useState('');
  const page = useRef(1);
  const pages = useRef(0);
  pages.current = data.pages;
  const movies = data.results;
  const endpoint = buildEndpoint(query, data.page);

  useEffect(() => {
    window.addEventListener('scroll', scrollEvent);

    return () => {
      window.removeEventListener('scroll', scrollEvent);
    };
  }, []);

  useEffect(() => {
    if (data.init && data.results.length === 0) {
      fetchMovies(data, endpoint, setData);
    }
  }, [data]);

  useEffect(() => {
    if (!data.init) {
      fetchMovies(data, endpoint, setData);
    }
  }, [page.current]);

  // Perform a cleanup after each new query
  useEffect(() => {
    if (!data.init) {
      page.current = 1;
      setData({
        init: true,
        page: 1,
        pages: 0,
        results: [],
      });
    }
  }, [query]);

  const scrollEvent = () => {
    const offsetHeight = document.body.offsetHeight;
    const innerHeight = window.innerHeight;
    const scrollY = window.scrollY;

    if (innerHeight + scrollY === offsetHeight) {
      if (page.current < pages.current) {
        page.current = page.current + 1;
        setData((state) => {
          const newState = Object.assign({}, state);
          newState.page = page.current;
          return newState;
        });
      }
    }
  };

  return (
    <MoviesContext.Provider value={{ movies, query, setQuery }}>
      {children}
    </MoviesContext.Provider>
  );
};

MoviesStateProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
