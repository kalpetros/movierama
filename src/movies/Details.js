import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Overview } from './Overview';
import { SimilarMovies } from './SimilarMovies';
import { Reviews } from './Reviews';

export const Details = (props) => {
  const [view, setView] = useState('overview');
  const { id: id, isSelected: isSelected } = props;

  const menuSelectedClass = 'movie__details__menu--selected';
  let baseClass = isSelected
    ? 'movie__details movie__details--active'
    : 'movie__details movie__details--inactive';
  let currentView = null;

  const handleChangeView = (e) => {
    const id = e.currentTarget.dataset.id;
    setView(id);
  };

  if (isSelected) {
    if (view === 'overview') {
      currentView = <Overview id={id} />;
    } else if (view === 'reviews') {
      currentView = <Reviews id={id} />;
    } else if (view === 'similar') {
      currentView = <SimilarMovies id={id} />;
    }
  }

  return (
    <div className={baseClass}>
      {currentView}
      <div className="movie__details__menu">
        <ul>
          <li
            data-id="overview"
            className={view === 'overview' ? menuSelectedClass : null}
            onClick={handleChangeView}
          >
            Overview
          </li>
          <li
            data-id="reviews"
            className={view === 'reviews' ? menuSelectedClass : null}
            onClick={handleChangeView}
          >
            Reviews
          </li>
          <li
            data-id="similar"
            className={view === 'similar' ? menuSelectedClass : null}
            onClick={handleChangeView}
          >
            Similar Movies
          </li>
        </ul>
      </div>
    </div>
  );
};

Details.defaultProps = {
  id: 1,
  isSelected: false,
};

Details.propTypes = {
  id: PropTypes.number.isRequired,
  isSelected: PropTypes.bool.isRequired,
};
