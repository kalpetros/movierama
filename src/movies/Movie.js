import React, { useContext, useState } from 'react';
import PropTypes from 'prop-types';

import { ConfigurationContext } from '../store/ConfigurationContext';
import { Details } from './Details';
import { GenreContext } from '../store/GenreContext';
import { buildImageUrl } from '../utils';

export const Movie = (props) => {
  const [isSelected, setIsSelected] = useState(false);
  const { state: configuration } = useContext(ConfigurationContext);
  const { state: genres } = useContext(GenreContext);
  const { data: data } = props;

  const handleClick = (e) => {
    setIsSelected((p) => (p ? false : true));
    const element = document.getElementById(`movie-${data.id}`);
    element.scrollIntoView({ behavior: 'smooth' });
  };

  let style = {};
  let overview = data.overview.length === 0 ? 'N/A' : data.overview;

  if (data.overview.length > 200) {
    overview = `${data.overview.substring(0, 200)}...`;
  }

  let legend = <p>{overview}</p>;
  const posterUrl = buildImageUrl(configuration, data.poster_path, 2);
  const year = new Date(data.release_date).getFullYear();

  let genresText = null;
  let genreList = data.genre_ids.map((id) => {
    const label = genres.find((g) => g.id === id);
    return label.name;
  });

  if (genreList.length > 0) {
    genresText = genreList.join(', ');
  }

  if (isSelected) {
    const popularity = Math.round(data.popularity, 0);
    const releaseDate = new Date(data.release_date);
    const date = releaseDate.getDate();
    const month = releaseDate.getMonth() + 1;
    const year = releaseDate.getFullYear();
    const dateString = `${date}/${month}/${year}`;
    const backgroundImage = buildImageUrl(configuration, data.poster_path, 5);

    legend = (
      <>
        <h3>Release date</h3>
        <p>{dateString}</p>
      </>
    );

    genresText = `Popularity - ${popularity}`;

    style = {
      backgroundImage: `url(${backgroundImage})`,
      backgroundRepeat: 'no-repeat',
      backgroundSize: 'cover',
    };
  }

  return (
    <div className="movie" style={style}>
      <div
        className="movie__wrapper"
        id={`movie-${data.id}`}
        onClick={handleClick}
      >
        <div className="movie__wrapper__content">
          <div className="movie__wrapper__image">
            <img src={posterUrl} alt="movie_poster" />
          </div>
          <div className="movie__wrapper__body">
            <div className="movie__wrapper__meta">
              <h2>
                {data.title} ({year})
              </h2>
              {legend}
            </div>
            <div className="movie__wrapper__footer">
              <div>
                <h3 className="font-bold">{genresText}</h3>
              </div>
              <div>
                <h3 className="font-bold text-right">
                  {data.vote_average} / 10
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
      <Details id={data.id} isSelected={isSelected} />
    </div>
  );
};

Movie.propTypes = {
  data: PropTypes.object.isRequired,
};
