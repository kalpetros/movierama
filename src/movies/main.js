import React, { useContext } from 'react';

import { MoviesContext } from '../store/movies/context';
import { Movie } from './Movie';

export const Movies = () => {
  const { movies } = useContext(MoviesContext);

  const list = movies.map((movie, index) => {
    return <Movie key={`movie-${index}`} data={movie} />;
  });

  return list;
};
