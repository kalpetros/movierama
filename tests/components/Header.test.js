import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

import { Header } from '../../src/components/Header';
import { MoviesContext } from '../../src/store/movies/context';

describe('<Header />', () => {
  let wrapper;

  it('should match the snapshot', () => {
    const mockContext = {
      movies: [],
      query: 'Test',
      setQuery: jest.fn(),
    };

    wrapper = mount(
      <MoviesContext.Provider value={mockContext}>
        <Header />
      </MoviesContext.Provider>
    );

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display movie title text', () => {
    const mockContext = {
      movies: [],
      query: 'Test',
      setQuery: jest.fn(),
    };

    wrapper = mount(
      <MoviesContext.Provider value={mockContext}>
        <Header />
      </MoviesContext.Provider>
    );

    const title = wrapper.find('.header__title').find('h1').text();
    expect(title).toBe('Movies "Test"');
  });

  it('should display in theaters text', () => {
    const mockContext = {
      movies: [],
      query: '',
      setQuery: jest.fn(),
    };

    wrapper = mount(
      <MoviesContext.Provider value={mockContext}>
        <Header />
      </MoviesContext.Provider>
    );

    const title = wrapper.find('.header__title').find('h1').text();
    expect(title).toBe('In Theaters');
  });
});
