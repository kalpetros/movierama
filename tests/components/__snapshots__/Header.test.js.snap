// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should match the snapshot 1`] = `"<header class=\\"header\\"><div class=\\"header__content\\"><div><div><h1>Welcome</h1></div><div><input type=\\"text\\" class=\\"input\\" placeholder=\\"Type here to search for your favorite movie...\\" value=\\"Test\\"></div></div></div><div class=\\"header__title\\"><div><h1>Movies \\"Test\\"</h1></div></div></header>"`;
